rfc-108-including-gem-component-assets.md

```diff
@@ -0,0 +1,189 @@
+# Include specific component assets in applications
+
+## Problem and background
+
+The [current method for including the components](https://github.com/alphagov/govuk_publishing_components/blob/master/docs/install-and-use.md#4-include-the-assets) in a frontend application (such as `government-frontend`) is to include `_all_components.scss` and `_all_components_print.scss` in the Sass of that application. This has some benefits:
+
+- all components are automatically included in the application
+
+This also has some detriments:
+
+- all components are included regardless of whether they are used, so unused CSS/JS may be included
+- only publically facing components can be added to the gem in order to prevent the previous problem
+
+We now have a number of private applications that use the components gem and components are being written for those applications. While the past model of putting components into applications has worked, we now have components that need to be shared across more than one private application, which is not possible with the current model without adding unnecessary page weight to www.gov.uk.
+
+Additionally, the way that the Sass is currently included in each application makes caching ineffective. When a user lands on a page on GOV.UK, their browser downloads the CSS for that application (say `government-frontend`). This is a single file that includes custom styles from that application plus the styles from the gem. If they then navigate to a page served by another application (say `collections`) they download a new CSS file that contains the exact same styles from the gem, but because it is mixed in with the custom styles for that application, they are different files and cannot be cached.
+
+This problem is potentially limited by gzipping but we could do with some numbers on this.
+
+## Possible solution 1
+
+Instead of including all of the component Sass, each application should be configured to only include the components that it needs. For example, in an applications `_application.scss`:
+
+```
+// currently
+@import 'govuk_publishing_components/all_components';
+```
+
+```
+// proposed
+@import 'govuk_publishing_components/components/accordion';
+@import 'govuk_publishing_components/components/button';
+// (etc.)
+```
+
+This would mean we can add any components we like to the gem without worrying that they are needlessly adding page weight to the publically facing GOV.UK.
+
+## Javascript
+
+We can do something similar with Javascript, however the initialisation of specific modules may be more complicated and require more thought. Additionally knowing whether or not to include jQuery could be difficult.
+
+As a MVP Javascript could be left as it is for now.
```

Comment:
I think this would make things confusing and will not help with reducing the JavaScript load on public facing/frontend application, which for me is more worrying than the CSS.

Reply:
I think it's worth mentioning that JavaScript could also be consumed individually by importing/requiring only the needed JavaScript files from govuk_publishing_components then initialise components individually or by using the `modules.js` initialiser.

Reply:
I've updated this line, I've left it deliberately vague as it seems like the specifics of this can be worked out during spikes - the overall intent is that JS files be consumed individually.